From 6e1bde9c1d54d647c7f4ae5f0650f8da257e2f38 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Fri, 19 Sep 2014 12:28:44 -0400
Subject: [pci] acpi: Use acpi_find_child_device() for child devices lookup

Message-id: <1411129776-21656-10-git-send-email-prarit@redhat.com>
Patchwork-id: 96345
O-Subject: [RHEL7.1 PATCH BZ 1141846 09/61] PCI / ACPI: Use acpi_find_child_device() for child devices lookup
Bugzilla: 1141846
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Myron Stowe <mstowe@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1141846

commit 5ce79d201358d36f13d13b01d8614bd8e646036c
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Thu Nov 28 23:58:08 2013 +0100

    PCI / ACPI: Use acpi_find_child_device() for child devices lookup

    It is much more efficient to use acpi_find_child_device()
    for child devices lookup in acpi_pci_find_device() and pass
    ACPI_COMPANION(dev->parent) to it directly instead of obtaining
    ACPI_HANDLE() of ACPI_COMPANION(dev->parent) and passing it to
    acpi_find_child() which has to run acpi_bus_get_device() to
    obtain ACPI_COMPANION(dev->parent) from that again.

    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
    Reviewed-by: Aaron Lu <aaron.lu@intel.com>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
Cc: Stefan Assmann <sassmann@redhat.com>
Cc: Myron Stowe <mstowe@redhat.com>
Cc: Andy Gospodarek <agospoda@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/acpi/glue.c b/drivers/acpi/glue.c
index 7c56c16..d854492 100644
--- a/drivers/acpi/glue.c
+++ b/drivers/acpi/glue.c
@@ -106,6 +106,9 @@ struct acpi_device *acpi_find_child_device(struct acpi_device *parent,
  struct acpi_device *adev, *ret = NULL;
  int ret_score = 0;
 
+ if (!parent)
+  return NULL;
+
  list_for_each_entry(adev, &parent->children, node) {
   unsigned long long addr;
   acpi_status status;
diff --git a/drivers/pci/pci-acpi.c b/drivers/pci/pci-acpi.c
index 988e75f..ae2c1fa 100644
--- a/drivers/pci/pci-acpi.c
+++ b/drivers/pci/pci-acpi.c
@@ -310,7 +310,8 @@ void acpi_pci_remove_bus(struct pci_bus *bus)
 static int acpi_pci_find_device(struct device *dev, acpi_handle *handle)
 {
  struct pci_dev *pci_dev = to_pci_dev(dev);
- bool is_bridge;
+ struct acpi_device *adev;
+ bool check_children;
  u64 addr;
 
  /*
@@ -318,14 +319,17 @@ static int acpi_pci_find_device(struct device *dev, acpi_handle *handle)
   * is set only after acpi_pci_find_device() has been called for the
   * given device.
   */
- is_bridge = pci_dev->hdr_type == PCI_HEADER_TYPE_BRIDGE
+ check_children = pci_dev->hdr_type == PCI_HEADER_TYPE_BRIDGE
    || pci_dev->hdr_type == PCI_HEADER_TYPE_CARDBUS;
  /* Please ref to ACPI spec for the syntax of _ADR */
  addr = (PCI_SLOT(pci_dev->devfn) << 16) | PCI_FUNC(pci_dev->devfn);
- *handle = acpi_find_child(ACPI_HANDLE(dev->parent), addr, is_bridge);
- if (!*handle)
-  return -ENODEV;
- return 0;
+ adev = acpi_find_child_device(ACPI_COMPANION(dev->parent), addr,
+          check_children);
+ if (adev) {
+  *handle = adev->handle;
+  return 0;
+ }
+ return -ENODEV;
 }
 
 static void pci_acpi_setup(struct device *dev)
-- 
1.7.1